render/detailed/node.go

```diff
@@ -0,0 +1,303 @@
+	report.Control
+}
+
+// MakeNode transforms a renderable node to a detailed node. It uses
+// aggregate metadata, plus the set of origin node IDs, to produce tables.
+func MakeNode(r report.Report, n render.RenderableNode) Node {
+	var hosts, pods, containerImages, containers, applications []ChildSummary
+	for id, child := range n.Children {
+		if id == n.ID {
+			continue
+		}
+
+		switch child.Topology {
+		case "process":
+			applications = append(applications, processChildSummary(child))
+		case "container":
+			containers = append(containers, containerChildSummary(child))
+		case "container_image":
+			containerImages = append(containerImages, containerImageChildSummary(child))
+		case "pod":
+			pods = append(pods, podChildSummary(child))
+		case "host":
+			hosts = append(hosts, hostChildSummary(child))
+		}
+	}
+	children := []ChildGroup{}
+	if len(hosts) > 0 {
+		children = append(children, ChildGroup{TopologyID: "hosts", Label: "Hosts", Nodes: hosts})
+	}
+	if len(pods) > 0 {
+		children = append(children, ChildGroup{TopologyID: "pods", Label: "Pods", Nodes: pods})
+	}
+	if len(containerImages) > 0 {
+		children = append(children, ChildGroup{TopologyID: "containers-by-image", Label: "Container Images", Nodes: containerImages})
+	}
+	if len(containers) > 0 {
+		children = append(children, ChildGroup{TopologyID: "containers", Label: "Containers", Nodes: containers})
+	}
+	if len(applications) > 0 {
+		children = append(children, ChildGroup{TopologyID: "applications", Label: "Applications", Nodes: applications})
+	}
+
+	return Node{
+		ID:       n.ID,
+		Label:    n.LabelMajor,
+		Rank:     n.Rank,
+		Pseudo:   n.Pseudo,
+		Controls: controls(r, n),
+		Metadata: NodeMetadata(r, n),
+		Metrics:  NodeMetrics(r, n),
+		Children: children,
+		Parents:  parents(r, n),
+	}
+}
+
+func controlsFor(topology report.Topology, nodeID string) []ControlInstance {
+	result := []ControlInstance{}
+	node, ok := topology.Nodes[nodeID]
+	if !ok {
+		return result
+	}
+
+	for _, id := range node.Controls.Controls {
+		if control, ok := topology.Controls[id]; ok {
+			result = append(result, ControlInstance{
+				ProbeID: node.Metadata[report.ProbeID],
+				NodeID:  nodeID,
+				Control: control,
+			})
+		}
+	}
+	return result
+}
+
+func controls(r report.Report, n render.RenderableNode) []ControlInstance {
+	if _, ok := r.Process.Nodes[n.ControlNode]; ok {
+		return controlsFor(r.Process, n.ControlNode)
+	} else if _, ok := r.Container.Nodes[n.ControlNode]; ok {
+		return controlsFor(r.Container, n.ControlNode)
+	} else if _, ok := r.ContainerImage.Nodes[n.ControlNode]; ok {
+		return controlsFor(r.ContainerImage, n.ControlNode)
+	} else if _, ok := r.Host.Nodes[n.ControlNode]; ok {
+		return controlsFor(r.Host, n.ControlNode)
+	}
+	return []ControlInstance{}
+}
+
+// parents is a total a hack to find the parents of a node (which is
+// ill-defined).
+func parents(r report.Report, n render.RenderableNode) (result []Parent) {
```

Comment:
Paul + I discussed this offline and Paul's going to try and generate this from the nodes themselves, instead of using the heuristics below.

The basic ideas is that a report.Node will have a Parents field, which will be a set of (topology, node id) tuples.  When you merge two report.Nodes (or render.RenderableNodes), this field will be unioned.

A render.RenderableNode (which embeds a report.Node) will add an extra field, Children, which will be automatically filled whenever a node is "Up" rendered, that is the "kind" of the node is changed (ie from endpoint to process).  When this happens, the current report.nodeID is added to the children, and node.parents is erased (to prevent eg processes having every host as a parent through the container image).

The observation is you should never merge two nodes of different kinds (ie an endpoint and a process node), instead you "up" merge some nodes (from endpoint to process), then merge the result with the process node.